public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/QueryEditor.tsx

```diff
@@ -142,28 +148,28 @@ function useLabels(
   useEffect(() => {
     const fetchData = async () => {
       const labels = await datasource.getLabelNames(
-        createSelector(query.labelSelector, query.profileTypeId, ''),
+        createSelector(queryRef.current.labelSelector, queryRef.current.profileTypeId, ''),
         unpreciseRange.from,
         unpreciseRange.to
       );
 
       setLabels(labels);
     };
     fetchData();
-  }, [query, unpreciseRange.from, unpreciseRange.to, datasource, setLabels]);
+  }, [queryRef, unpreciseRange.from, unpreciseRange.to, datasource, setLabels]);
 
   // Create a function with range and query already baked in, so we don't have to send those everywhere
   const getLabelValues = useCallback(
     (label: string) => {
-      let labelSelector = createSelector(query.labelSelector, query.profileTypeId, label);
+      const labelSelector = createSelector(queryRef.current.labelSelector, queryRef.current.profileTypeId, label);
       return datasource.getLabelValues(labelSelector, label, unpreciseRange.from, unpreciseRange.to);
     },
-    [datasource, query.labelSelector, query.profileTypeId, unpreciseRange.to, unpreciseRange.from]
+    [queryRef, datasource, unpreciseRange.from, unpreciseRange.to]
   );
 
   const onChangeDebounced = debounce((value: string) => {
     if (onChange) {
-      onChange({ ...query, labelSelector: value });
+      onChange({ ...queryRef.current, labelSelector: value });
```

Comment:
I think this is the root cause, because we don't update the query in here. So we can either minimize the code change and only use a ref for this without needing it in other places while keeping the the hook signature bit cleaner or we can change the usage a to pass the query to the debounce function when used.

Also it's not clear to me why these `onChangeDebounced` and `onLabelSelectorChange` are lumped together in this single hook if they don't need anything from the other code.

Would probably opt to separate them to their own hook and use the queryRef just inside of it that way I think it would be much clearer why the ref is needed.

Reply:
Ahh yes this is much cleaner to use a separate hook. I appreciate your detailed comment. Not 100% sure I addressed it all but at least the purpose of the ref is a lot more clear now without providing a ref in places it is unnecessary.